gitoid/src/gitoid.rs

```diff
@@ -181,30 +175,36 @@ impl TryFrom<Url> for GitOid {
         };
 
         // Parse the hash algorithm, if present.
-        let hash_algorithm = {
-            let part = segments
-                .next()
-                .and_then(|p| p.is_empty().not().then_some(p))
-                .ok_or_else(|| MissingHashAlgorithm(url.clone()))?;
+        let hash_algorithm = segments
+            .next()
+            .and_then(|p| p.is_empty().not().then_some(p))
+            .ok_or_else(|| MissingHashAlgorithm(url.clone()))?;
 
-            HashAlgorithm::from_str(part)?
-        };
+        if hash_algorithm != D::NAME {
+            return Err(Error::MismatchedHashAlgorithm(
+                D::NAME.to_string(),
+                hash_algorithm.to_string(),
+            ));
+        }
 
         // Parse the hash, if present.
         let hex_str = segments
             .next()
             .and_then(|p| p.is_empty().not().then_some(p))
             .ok_or_else(|| MissingHash(url.clone()))?;
-        let mut value = [0u8; 32];
+
+        let mut value = GenericArray::generate(|_| 0);
+
         hex::decode_to_slice(hex_str, &mut value)?;
 
+        let expected_size = <D as OutputSizeUser>::output_size();
+
+        if value.len() != expected_size {
+            return Err(Error::UnexpectedHashLength(expected_size, value.len()));
```

Comment:
By convention, I'm putting expected first and observed second. These _could_ be replaced with named fields instead for clarity.